### PR TITLE
IntroduceLocal on simple names

### PR DESCRIPTION
**Pull request by jcouv**
### Customer scenario
Allow introducing a local for simple expressions.

### Bugs this fixes
Fixes https://github.com/dotnet/roslyn/issues/10123

### Is this a regression from a previous update?
No

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 1]
```diff
@@ -4616,6 +4616,173 @@ void M()
...
+        [WorkItem(10123, "https://github.com/dotnet/roslyn/issues/10123")]
+        public async Task TestSimpleParameterName()
+        {
+            await TestInRegularAndScriptAsync(
+@"
+class C
+{
+    void M(int a)
+    {
+        System.Console.Write([|a|]);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs]: Does thsi mean on every local/parameter/variable, i'll always get an optin to make a local out of it?

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs [hunk 2]
```diff
@@ -229,14 +229,21 @@ private TExpressionSyntax GetExpressionUnderSpan(SyntaxTree tree, TextSpan textS
...
             {
                 if (!_service.CanIntroduceVariableFor(this.Expression))
                 {
                     return false;
                 }
 
-                if (this.Expression is TTypeSyntax)
+                if (isSpanEmpty && this.Expression is TNameSyntax)
+                {
+                    // to extract a name, you must have a selection (this avoids making the refactoring too noisy)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs]: ah, nice.  that seems like a good thing to do.

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs [hunk 3]
```diff
@@ -229,14 +229,21 @@ private TExpressionSyntax GetExpressionUnderSpan(SyntaxTree tree, TextSpan textS
...
+                    return false;
+                }
+
+                if (this.Expression is TTypeSyntax && !(this.Expression is TNameSyntax))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs]: probably good to comment this.

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs [hunk 4]
```diff
@@ -18,12 +18,13 @@
...
 {
-    internal abstract partial class AbstractIntroduceVariableService<TService, TExpressionSyntax, TTypeSyntax, TTypeDeclarationSyntax, TQueryExpressionSyntax> : IIntroduceVariableService
-        where TService : AbstractIntroduceVariableService<TService, TExpressionSyntax, TTypeSyntax, TTypeDeclarationSyntax, TQueryExpressionSyntax>
+    internal abstract partial class AbstractIntroduceVariableService<TService, TExpressionSyntax, TTypeSyntax, TTypeDeclarationSyntax, TQueryExpressionSyntax, TNameSyntax> : IIntroduceVariableService
+        where TService : AbstractIntroduceVariableService<TService, TExpressionSyntax, TTypeSyntax, TTypeDeclarationSyntax, TQueryExpressionSyntax, TNameSyntax>
         where TExpressionSyntax : SyntaxNode
         where TTypeSyntax : TExpressionSyntax
         where TTypeDeclarationSyntax : SyntaxNode
         where TQueryExpressionSyntax : TExpressionSyntax
+        where TNameSyntax : SyntaxNode
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs]: Note: TNameSyntax should probably derive from TTypeSyntax

**Comment by jcouv** [comment]
@dotnet/roslyn-ide for review for 15.7. Thanks

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 5]
```diff
@@ -4616,6 +4616,173 @@ void M()
...
+    }
+}",
+@"
+class C
+{
+    void M(int a)
+    {
+        int {|Rename:a1|} = a;
```
Comment (heejaechang) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs]: test that prefer var?

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs [hunk 6]
```diff
@@ -229,15 +229,23 @@ private TExpressionSyntax GetExpressionUnderSpan(SyntaxTree tree, TextSpan textS
...
+                bool isSpanEmpty,
                 CancellationToken cancellationToken)
             {
                 if (!_service.CanIntroduceVariableFor(this.Expression))
                 {
                     return false;
                 }
 
-                if (this.Expression is TTypeSyntax)
+                if (isSpanEmpty && this.Expression is TNameSyntax)
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs]: When the expression is found for a given span, there is a special case for empty span. That's why it needs to be treated specially here.

If you place your cursor on a name, the expression will be found, and it must be rejected (should not introduce variable).
If you select a portion of a name, the expression is not found.
If you select a whole name, the expression is found and it is accepted (can introduce variable).

There are tests to illustrate those scenarios (`_EmptySelection`, `_SmallSelection`).

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs (hunk 3 above)
Comment (heejaechang) [reply src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs]: so condition that can hit here is

1, span not empty and expression is not name syntax 
2, span is empty but expression is not name syntax 
3. span not empty but expression is name syntax 

so, if expression is not name syntax, regardless of span, if expression is typesyntax, we don't support refactoring.

shouldn't this check move before isSpanEmpty check?

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs (hunk 5 above)
Comment (heejaechang) [reply src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs]: ignore this. I see this use existing introduce var code. which should take care of code style.

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (heejaechang) [reply src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs]: I see. it relies on the fact that FindNode will not return expression if selection doesn't include whole expression. 

yep, I saw the test. and wondered where that rejection happens. 

thank you.

**Comment by jcouv** [comment]
@jinujoseph for ask-mode approval for 15.7. Or let me know if I should push out to 15.8. Thanks

**Comment by jinujoseph** [comment]
@Pilchie for 15.7.preview4 ask mode approval

**Comment by Pilchie** [comment]
Approved for 15.7 P4.

**Comment by DustinCampbell** [comment]
This PR makes me *so* happy. It's always been a frustration of mine that I can't use Introduce Local in more cases.

**Pull request merged**